syntax = "proto3";

package common.gossip.rpc;

import "google/protobuf/empty.proto";

option cc_generic_services = true;

enum State {
  ALIVE = 0;
  SUSPECT = 1;
  DEAD = 2;
  LEFT = 3;
}

message Node {
  string name = 1;
  string ip = 2;
  uint32 port = 3;
  State state = 4;
  bytes metadata = 5;
}

message PingReq {
  string ip = 1;
  uint32 port = 2;
}

message SuspectMsg {
  uint32 version = 1;
  string dst = 2;
  repeated string srcs = 3;
}

message AckResp {
  enum Type {
    ACK = 0;
    NACK = 1;
  }
  Type type = 2;
}

message AliveMsg {
  uint32 version = 1;
  string name = 2;
  string ip = 3;
  uint32 port = 4;
  bytes metadata = 5;
}

message DeadMsg {
  uint32 version = 1;
  string name = 2;
  repeated string from = 3;
}

message SyncMsg {
  uint32 version = 1;
  repeated Node nodes = 2;
}

service ServerAPI {
  rpc Ping(google.protobuf.Empty) returns (google.protobuf.Empty) {}
  rpc IndirectPing(PingReq) returns (AckResp) {}
  rpc Suspect(SuspectMsg) returns (google.protobuf.Empty) {}
  rpc Sync(SyncMsg) returns (SyncMsg) {}
  rpc Dead(DeadMsg) returns (google.protobuf.Empty) {}
}
